unit/util/piped_process.cpp

```diff
@@ -0,0 +1,256 @@
+/// \file
+/// \author Diffblue Ltd.
+/// Unit tests for checking the piped process communication mechanism.
+
+#ifdef _WIN32
+// No unit tests yet!
+#else
+
+#  include <testing-utils/use_catch.h>
+#  include <util/piped_process.h>
+
+TEST_CASE(
+  "We create a pipe and we can read from it",
+  "[core][util][piped_process]")
+{
+  const std::string to_be_echoed = "The Jabberwocky";
+  // Need to give path to avoid shell built-in invocation
+  const std::string binary = "/bin/echo";
+  const std::string command = binary + " " + to_be_echoed;
+  piped_processt process = piped_processt(command);
+
+  // This is an indirect way to detect when the pipe has something since
+  // -1 is an infinite timeout wait. This could (in theory) also return
+  // when there is an error, but this unit test is not doing error handling.
+  process.can_receive(-1);
+  std::string response = process.receive();
+
+  response.erase(
+    std::remove(response.begin(), response.end(), '\n'), response.end());
+
+  REQUIRE(response == to_be_echoed);
+}
+
+TEST_CASE(
+  "We create a pipe, send and receive from it",
+  "[core][util][piped_process]")
+{
+  const std::string binary = "z3 -in";
+  const std::string statement = "(echo \"hi\")\n";
+  const std::string termination_statement = "(exit)\n";
+  piped_processt process = piped_processt(binary);
+
+  piped_processt::process_send_responset res = process.send(statement);
+  REQUIRE(res == piped_processt::process_send_responset::SUCCEEDED);
```

Comment:
:bulb: If you inline `process.send(statement)` into this requirement, then catch will give more informative error messages in the case where this fails. This is because the error messaging will show `process.send(statement)` on that side of the equals instead of just `res`. `REQUIRE(process.send(statement) == piped_processt::process_send_responset::SUCCEEDED);`